packages/core/src/routes/admin-user.ts

```diff
@@ -0,0 +1,43 @@
+import { userInfoSelectFields } from '@logto/schemas';
+import pick from 'lodash.pick';
+import { InvalidInputError } from 'slonik';
+import { object, string } from 'zod';
+
+import koaGuard from '@/middleware/koa-guard';
+import { findRolesByRoleName } from '@/queries/roles';
+import { findUserById, updateUserById } from '@/queries/user';
+
+import { AuthedRouter } from './types';
+
+export default function adminUserRoutes<T extends AuthedRouter>(router: T) {
+  router.patch(
+    '/users/:userId/roles',
+    koaGuard({
+      params: object({ userId: string().min(1) }),
+      body: object({ roleNames: string().array().nullable() }),
```

Comment:
should be non-nullable since we use `[]` for explicitly setting empty?

Reply:
1. `roles` -> `roleNames` to reflect actual status?
2. also `roleNames` -> `names` since it's already under `roleNames` route

Reply:
`roleNames` in DB is nullable.  Considering roles could be an optional field, if it is never set to a user, the user data should not return the roleNames field.

Reply:
Add a new role at the `convertToPrimitiveOrSql`: if is empty array return null;